eth/vm/gas_meter.py

```diff
@@ -16,17 +17,49 @@
 )
 
 
+def default_refund_strategy(meter: "GasMeter", amount: int) -> None:
+    if amount < 0:
+        raise ValidationError("Gas refund amount must be positive")
+
+    meter.gas_refunded += amount
+
+    meter.logger.trace(
+        'GAS REFUND: %s + %s -> %s',
+        meter.gas_refunded - amount,
+        amount,
+        meter.gas_refunded,
+    )
+
+
+def allow_negative_refund_strategy(meter: "GasMeter", amount: int) -> None:
+    meter.gas_refunded += amount
+
+    meter.logger.trace(
+        'GAS REFUND: %s + %s -> %s',
+        meter.gas_refunded - amount,
+        amount,
+        meter.gas_refunded,
+    )
+
+
+RefundStrategy = Callable[["GasMeter", int], None]
+
+
 class GasMeter(object):
+
     start_gas = None  # type: int
 
     gas_refunded = None  # type: int
     gas_remaining = None  # type: int
 
     logger = cast(TraceLogger, logging.getLogger('eth.gas.GasMeter'))
 
-    def __init__(self, start_gas: int) -> None:
+    def __init__(self,
+                 start_gas: int,
+                 refund_strategy: RefundStrategy = default_refund_strategy) -> None:
```

Comment:
@pipermerriam @carver I wanted to avoid `ConstantinopleGasMeter` and reached for a more compositional approach that passes in a `RefundStrategy`. I don't have strong feelings about it. If you rather like this to follow the path that the rest of the code base uses (introducing `ConstantinopleGasMeter` and overwriting `refund_gas`) that is cool with me as well. In *general*, I'm more a friend of the compositional approach, it's just that this change is so tiny that I think there isn't much harm from an inheritance approach either.

Reply:
Yeah, seems promising, and a reasonable place to experiment with the approach because it's so simple. :+1: